File: play-services-core/src/main/java/org/microg/gms/asterism/AsterismServiceImpl.java
/*
 * Copyright (C) 2013-2017 microG Project Team
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.microg.gms.asterism;

import android.content.Context;
import android.content.SharedPreferences;
import android.os.RemoteException;
import android.util.Log;

import com.google.android.gms.asterism.GetAsterismConsentRequest;
import com.google.android.gms.asterism.internal.IAsterismApiService;
import com.google.android.gms.common.api.Status;
import com.google.android.gms.common.api.internal.IStatusCallback;

public class AsterismServiceImpl extends IAsterismApiService.Stub {
    private static final String TAG = "GmsAsterismSvc";
    private static final String PREF_NAME = "asterism_consent";
    private static final String PERMISSION_RCS = "com.google.android.gms.permission.RCS";

    private final Context context;

    public AsterismServiceImpl(Context context) {
        this.context = context;
    }

    @Override
    public void getAsterismConsent(IStatusCallback callback, GetAsterismConsentRequest request) throws RemoteException {
        if (callback == null) {
            Log.w(TAG, "getAsterismConsent: callback is null");
            return;
        }
        try {
            if (context.checkCallingPermission(PERMISSION_RCS) != android.content.pm.PackageManager.PERMISSION_GRANTED) {
                Log.w(TAG, "Missing " + PERMISSION_RCS);
                try {
                    callback.onResult(Status.INTERNAL_ERROR);
                } catch (RemoteException re) {
                    Log.w(TAG, "Failed to deliver permission error callback", re);
                }
                return;
            }

            String account = (request != null && request.accountName != null) ? request.accountName : "unknown";
            Log.d(TAG, "getAsterismConsent for: " + redact(account));

            SharedPreferences prefs = context.getSharedPreferences(PREF_NAME, Context.MODE_PRIVATE);
            boolean consented = prefs.getBoolean(account, true);
            try {
                callback.onResult(consented ? Status.SUCCESS : Status.CANCELED);
            } catch (RemoteException re) {
                Log.w(TAG, "Failed to deliver consent result", re);
            }
        } catch (Exception e) {
            Log.w(TAG, "getAsterismConsent failed defensively; replying SUCCESS to avoid caller deadlock", e);
            try {
                callback.onResult(Status.SUCCESS);
            } catch (RemoteException re) {
                Log.w(TAG, "Failed to deliver defensive SUCCESS callback", re);
            }
        }
    }

    private String redact(String account) {
        if (account == null || account.length() <= 4) return "***";
        return account.substring(0, 2) + "***" + account.substring(account.length() - 2);
    }
}

File: play-services-core/src/main/kotlin/org/microg/gms/asterism/AsterismServiceImpl.kt
/*
 * SPDX-FileCopyrightText: 2026 microG Project Team
 * SPDX-License-Identifier: Apache-2.0
 */

package org.microg.gms.asterism

import android.content.Context
import android.content.SharedPreferences
import android.util.Log
import com.google.android.gms.asterism.*
import com.google.android.gms.asterism.internal.IAsterismApiService
import com.google.android.gms.asterism.internal.IAsterismCallbacks
import com.google.android.gms.common.api.Status
import java.security.MessageDigest
import java.util.UUID

private const val TAG = "GmsAsterism"
private const val PREFS_NAME = "asterism_consent"
private const val KEY_CONSENTED = "consented"
private const val KEY_CONSENT_TOKEN = "consent_token"

class AsterismServiceImpl(private val context: Context) : IAsterismApiService.Stub() {

    private val prefs: SharedPreferences by lazy {
        context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)
    }

    override fun getAsterismConsent(callbacks: IAsterismCallbacks?, request: GetAsterismConsentRequest?) {
        Log.d(TAG, "getAsterismConsent($request)")
        val response = GetAsterismConsentResponse()
        response.consented = prefs.getBoolean(KEY_CONSENTED, true)
        response.consentToken = getOrCreateConsentToken()
        callbacks?.onGetAsterismConsent(Status.SUCCESS, response)
    }

    override fun setAsterismConsent(callbacks: IAsterismCallbacks?, request: SetAsterismConsentRequest?) {
        Log.d(TAG, "setAsterismConsent(consented=${request?.consented})")
        val response = SetAsterismConsentResponse()
        if (request != null) {
            prefs.edit()
                .putBoolean(KEY_CONSENTED, request.consented)
                .putString(KEY_CONSENT_TOKEN, generateConsentToken())
                .apply()
            response.success = true
        } else {
            response.success = false
        }
        callbacks?.onSetAsterismConsent(if (response.success) Status.SUCCESS else Status.ERROR, response)
    }

    private fun getOrCreateConsentToken(): String {
        val existing = prefs.getString(KEY_CONSENT_TOKEN, null)
        if (existing != null) return existing
        val token = generateConsentToken()
        prefs.edit().putString(KEY_CONSENT_TOKEN, token).apply()
        return token
    }

    private fun generateConsentToken(): String {
        return try {
            val source = arrayOf(
                android.os.Build.FINGERPRINT,
                android.os.Build.SERIAL,
                context.packageName
            ).joinToString("|")

            val digest = MessageDigest.getInstance("SHA-256").digest(source.toByteArray())
            digest.joinToString("") { "%02x".format(it) }.take(32)
        } catch (e: Exception) {
            Log.w(TAG, "Failed to generate stable consent token, using UUID", e)
            UUID.randomUUID().toString().replace("-", "").take(32)
        }
    }
}
